lib/cn/cn_tree.c

```diff
@@ -2211,6 +2211,18 @@ cn_tree_node_mclass(struct cn_tree_node *tn, enum hse_mclass_policy_dtype dtype)
     return mclass_policy_get_type(policy, age, dtype);
 }
 
+uint
+cn_tree_node_scatter(const struct cn_tree_node *tn)
+{
+    struct kvset_list_entry *le;
+    uint scatter = 0;
+
+    list_for_each_entry(le, &tn->tn_kvset_list, le_link)
+        scatter += kvset_get_scatter(le->le_kvset, scatter);
+
+    return scatter;
+}
```

Comment:
A node with 8 kvsets, each with 1 vgroup as scatter == 0.
If the oldest ksvet has 2 vgroups, and all others have 1, then the node scatter is 2.
If the newest kvset has 2 vgroups, and all others have 1, the node scatter is 10.
That doesn't seem right.

Reply:
Why not simply define node scatter as sum of vgroups - number of kvsets ?  Then the above examples would have scatter of 0, 1 and 1.

Reply:
I think maybe you're assuming kvset_get_scatter() simply returns the vgroup count, right?  It's not that simplistic...

If oldest kvset has 2 vgroups and all others have 1 then the scatter is 9.
If the newest kvset has 2 vgroups and all other have 1 then scatter is 2.

Searching from the oldest to youngest kvsets, we skip fully compacted kvsets (i.e., vgroups === 1) until we find a kvset with (vgroups > 1), then we sum up all the kvsets of the remaining younger kvsets to get the scatter score.
Prior to patch I didn't sum up the younger kvsets if they were fully compacted.  But now I do so that when the leaf-len rule and scatter rule both see that the scatter score reaches the HWM then the leaf-len rule will defer to the scatter rule until the leaf-len gets excessively long then it will do some smaller fixed k-compacts while waiting for a scatter job to come along.

The scatter rule does try to avoid including the youngest kvsets with no scatter, but it will take all such kvsets above the lrunlen_min threshold (4 kvsets).  We could include them as well, but seems preferable not to turn this job into a full gc operation...

Reply:
Right.  I just had it backwards and was off by one.  My point is that it seems odd that the order of kvsets would affect the nodes scatter metric.

Reply:
Now that you made me think about it, there's another k-compaction optimization we can make.  We've long had this problem (looking at compc) where something like 0 0 0 1 1 1 2 2 2 will require three k-compacts in a row to reduce once we add one more new kvset to the left of the list.

But I just realized I can do that in one k-compact with this one simple trick!  (i.e., only reset the mark if the current run length is less than (runlen_min - 1).  Woot!

Reply:
With this one simple trick!